File: Sources/LibTorch/tensor_factory.cpp
#include <torch/torch.h>
#include "tensor_factory.h"

using std::swap;

Tensor* arange(long start, long end, long step, TensorOptions* opt) {
  auto c10opt = (c10::TensorOptions*) opt;
  auto t = torch::arange(start, end, step, *c10opt);
  torch::Tensor* p = new torch::Tensor();
  swap(t, *p);
  return (Tensor*) p;
}

Tensor* zeros(long* sizes, uint64_t length, TensorOptions* opt) {
  auto c10opt = (c10::TensorOptions*) opt;
  auto t = torch::zeros(c10::IntArrayRef(sizes, length), *c10opt);
  torch::Tensor* p = new torch::Tensor();
  swap(t, *p);
  return (Tensor*) p;
}

Tensor* ones(long* sizes, uint64_t length, TensorOptions* opt) {
  auto c10opt = (c10::TensorOptions*) opt;
  auto t = torch::ones(c10::IntArrayRef(sizes, length), *c10opt);
  torch::Tensor* p = new torch::Tensor();
  swap(t, *p);
  return (Tensor*) p;
}

Tensor* randn(long* sizes, uint64_t length, TensorOptions* opt) {
  auto c10opt = (c10::TensorOptions*) opt;
  auto t = torch::randn(c10::IntArrayRef(sizes, length), *c10opt);
  torch::Tensor *p = new torch::Tensor();
  swap(t, *p);
  return (Tensor*) p;
}

Tensor* randint(int64_t low, int64_t high, long* sizes, uint64_t length, TensorOptions* opt) {
  auto c10opt = (c10::TensorOptions*) opt;
  auto t = torch::randint(low, high, c10::IntArrayRef(sizes, length), *c10opt);
  torch::Tensor *p = new torch::Tensor();
  swap(t, *p);
  return (Tensor*) p;
}

File: Sources/Switch/TensorFactory.swift
import LibTorch


func arange(_ start:Int, _ end:Int, _ step:Int, dtype:String="float") -> Tensor {
    let opt:TensorOptions = TensorOptionsFromDtype(dtype);
    return Tensor(LibTorch.arange(start, end, step, opt._p))
}

func zeros(_ sizes:Int..., dtype:String="float") -> Tensor {
    var sizes64 = sizes;
    let opt:TensorOptions = TensorOptionsFromDtype(dtype);
    return Tensor(LibTorch.zeros(&sizes64, UInt64(sizes.count), opt._p))
}

func ones(_ sizes:Int..., dtype:String="float") -> Tensor {
    var sizes64 = sizes;
    let opt:TensorOptions = TensorOptionsFromDtype(dtype);
    return Tensor(LibTorch.ones(&sizes64, UInt64(sizes.count), opt._p))
}

func randn(_ sizes:Int..., dtype:String="float") -> Tensor {
    var sizes64 = sizes;
    let opt:TensorOptions = TensorOptionsFromDtype(dtype);
    return Tensor(LibTorch.randn(&sizes64, UInt64(sizes.count), opt._p))
}

func randint(_ low:Int, _ high:Int, _ sizes:Int..., dtype:String="float") -> Tensor {
    var sizes64 = sizes;
    let opt:TensorOptions = TensorOptionsFromDtype(dtype);
    return Tensor(LibTorch.randint(Int64(low), Int64(high), &sizes64, UInt64(sizes.count), opt._p))
}
